src_lockfree/spsc_queue.ml

```diff
@@ -24,82 +24,79 @@
 type 'a t = {
   array : 'a Option.t Array.t;
   tail : int Atomic.t;
+  tail_cache : int ref;
   head : int Atomic.t;
-  mask : int;
+  head_cache : int ref;
 }
 
 exception Full
 
 let create ~size_exponent =
-  let size = Int.shift_left 1 size_exponent in
-  {
-    head = Atomic.make 0;
-    tail = Atomic.make 0;
-    mask = size - 1;
-    array = Array.init size (fun _ -> None);
-  }
+  if size_exponent < 0 || Sys.int_size - 2 < size_exponent then
+    invalid_arg "size_exponent out of range";
+  let size = 1 lsl size_exponent in
+  let array = Array.make size None in
+  let tail = Atomic.make_contended 0 in
+  let tail_cache = ref 0 in
+  let head = Atomic.make_contended 0 in
+  let head_cache = ref 0 in
```

Comment:
The cached index refs, `tail_cache` and `head_cache`, will potentially suffer from false sharing.

The cache indices do not need to be stored as atomics, but you could change these to be contended atomics, for example.  It would likely improve performance under contention.  (I actually checked and this improved performance on my laptop.) Alternatively you could otherwise allocated the cached indices to avoid false sharing.  For example, if you would allocate a `int array` of correct size,

```ocaml
let tail_cache = Array.make (Obj.size (Obj.repr tail)) 0
```

then that would avoid the false sharing with the current runtime.  (You would then use the index 0 of the array as the cached index value.)